src/main/java/org/engine/rest/UsersController.java

```diff
@@ -59,50 +65,210 @@
 
-        return null;
+        return userService.findByLogin(resetUserDTO.getName()).map(user -> {
+
+            if (!user.getEmail().equals(resetUserDTO.getEmail())) {
+                return new ResponseEntity<>("NAME_AND_EMAIL_MISMATCH", HttpStatus.BAD_REQUEST);
+            } else {
+
+                user.setResetPasswordTokenSentAt(LocalDateTime.now());
+                userService.save(user);
+
+                resetHandler.sendResetMail(user);
+            }
+
+            return ok(resetUserDTO);
+        })
+                .orElseGet(() -> notFound().build());
     }
 
+    // Step 2 - user opens link from e-mail
+
+    /**
+     * Called when link from reset e-mail is opened
+     * @param resetPasswordTokenDTO
+     * @return
+     */
     @PostMapping("reset_token")
-    public ResponseEntity<?> reset_token(@RequestBody AuthenticationDTO resetDTO) {
+    public ResponseEntity<?> resetToken(@Valid @RequestBody ResetPasswordTokenDTO resetPasswordTokenDTO) {
 
-        return null;
-    }
+        return userService.findByResetPasswordToken(resetPasswordTokenDTO.getResetPasswordToken()).map(user -> {
 
-    @PostMapping("confirmation_token")
-    public ResponseEntity<?> confirmation_token(@RequestBody AuthenticationDTO resetDTO) {
+            // We have a window of 24 hours to open the reset link from e-mail. If it's old return not found
+            long hours = ChronoUnit.HOURS.between(user.getConfirmationSentAt(), LocalDateTime.now());
 
-        return null;
+            // TODO - add logic to get the time also from the token and make a check
+
+            if(hours <= 24)
+            {
+                user.setResetPasswordToken(null);
+                resetPasswordTokenDTO.setStatus(HttpStatus.OK.value()); // Return status 200
+                resetPasswordTokenDTO.setLogin(user.getLogin());
+
+                userService.save(user);
+
+                return ok(resetPasswordTokenDTO);
+            }
+
+            return notFound().build();
+        }).orElseGet(() -> notFound().build());
     }
 
+    // Step 3 - user submits new password using web form
+
+    /**
+     * Called when server sends status 200 for token confirmation and form is submitted
+     * @param resetPasswordDTO
+     * @return
+     */
     @PostMapping("reset_password")
-    public ResponseEntity<?> reset(@RequestBody AuthenticationDTO resetDTO) {
+    public ResponseEntity<?> resetPassword(@Valid @RequestBody ResetPasswordDTO resetPasswordDTO) {
 
-        return null;
+        // TODO - think how to get the name - we delete ResetPasswordToken from DB in step 2
+
+        if(!jwtTokenUtil.validateToken(resetPasswordDTO.getResetPasswordToken(), new JwtUser(resetPasswordDTO.getLogin())))
+        {
+            return new ResponseEntity<>("INVALID_TOKEN", HttpStatus.BAD_REQUEST);
+        }
+        else
+        {
+            return this.userService.findByLogin(resetPasswordDTO.getLogin()).map(user -> {
```

Comment:
Again, Business logic here making the Controller unreadable
The controller should just be focus in received requests, delegating tasks, and handling responses